File: com/pdfjet/Box.java
/**
 *  Box.java
 *
Copyright 2020 Innovatics Inc.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and / or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

package com.pdfjet;


/**
 *  Used to create rectangular boxes on a page.
 *  Also used to for layout purposes. See the placeIn method in the Image and TextLine classes.
 *
 */
public class Box implements Drawable {

    protected float x;
    protected float y;

    private float w;
    private float h;

    private int color = Color.black;

    private float width = 0.3f;
    private String pattern = "[] 0";
    private boolean fillShape = false;

    private String language = null;
    private String altDescription = Single.space;
    private String actualText = Single.space;

    protected String uri = null;
    protected String key = null;


    /**
     *  The default constructor.
     *
     */
    public Box() {
    }


    /**
     *  Creates a box object.
     *
     *  @param x the x coordinate of the top left corner of this box when drawn on the page.
     *  @param y the y coordinate of the top left corner of this box when drawn on the page.
     *  @param w the width of this box.
     *  @param h the height of this box.
     */
    public Box(double x, double y, double w, double h) {
        this.x = (float) x;
        this.y = (float) y;
        this.w = (float) w;
        this.h = (float) h;
    }


    /**
     *  Creates a box object.
     *
     *  @param x the x coordinate of the top left corner of this box when drawn on the page.
     *  @param y the y coordinate of the top left corner of this box when drawn on the page.
     *  @param w the width of this box.
     *  @param h the height of this box.
     */
    public Box(float x, float y, float w, float h) {
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;
    }

    /**
     *  Sets the position of this box on the page.
     *
     *  @param x the x coordinate of the top left corner of this box when drawn on the page.
     *  @param y the y coordinate of the top left corner of this box when drawn on the page.
     */
    public void setPosition(float x, float y) {
        setLocation(x, y);
    }


    /**
     *  Sets the position of this box on the page.
     *
     *  @param x the x coordinate of the top left corner of this box when drawn on the page.
     *  @param y the y coordinate of the top left corner of this box when drawn on the page.
     */
    public void setPosition(double x, double y) {
        setLocation(x, y);
    }


    /**
     *  Sets the location of this box on the page.
     *
     *  @param x the x coordinate of the top left corner of this box when drawn on the page.
     *  @param y the y coordinate of the top left corner of this box when drawn on the page.
     *  @return this Box object.
     */
    public Box setLocation(float x, float y) {
        this.x = x;
        this.y = y;
        return this;
    }


    /**
     *  Sets the location of this box on the page.
     *
     *  @param x the x coordinate of the top left corner of this box when drawn on the page.
     *  @param y the y coordinate of the top left corner of this box when drawn on the page.
     *  @return this Box object.
     */
    public Box setLocation(double x, double y) {
        return setLocation((float) x, (float) y);
    }


    /**
     *  Sets the size of this box.
     *
     *  @param w the width of this box.
     *  @param h the height of this box.
     */
    public void setSize(double w, double h) {
        this.w = (float) w;
        this.h = (float) h;
    }


    /**
     *  Sets the size of this box.
     *
     *  @param w the width of this box.
     *  @param h the height of this box.
     */
    public void setSize(float w, float h) {
        this.w = w;
        this.h = h;
    }


    /**
     *  Sets the color for this box.
     *
     *  @param color the color specified as an integer.
     */
    public void setColor(int color) {
        this.color = color;
    }


    /**
     *  Sets the width of this line.
     *
     *  @param width the width.
     */
    public void setLineWidth(double width) {
        this.width = (float) width;
    }


    /**
     *  Sets the width of this line.
     *
     *  @param width the width.
     */
    public void setLineWidth(float width) {
        this.width = width;
    }


    /**
     *  Sets the URI for the "click box" action.
     *
     *  @param uri the URI
     */
    public void setURIAction(String uri) {
        this.uri = uri;
    }


    /**
     *  Sets the destination key for the action.
     *
     *  @param key the destination name.
     */
    public void setGoToAction(String key) {
        this.key = key;
    }


    /**
     *  Sets the alternate description of this box.
     *
     *  @param altDescription the alternate description of the box.
     *  @return this Box.
     */
    public Box setAltDescription(String altDescription) {
        this.altDescription = altDescription;
        return this;
    }


    /**
     *  Sets the actual text for this box.
     *
     *  @param actualText the actual text for the box.
     *  @return this Box.
     */
    public Box setActualText(String actualText) {
        this.actualText = actualText;
        return this;
    }


    /**
     *  The line dash pattern controls the pattern of dashes and gaps used to stroke paths.
     *  It is specified by a dash array and a dash phase.
     *  The elements of the dash array are positive numbers that specify the lengths of
     *  alternating dashes and gaps.
     *  The dash phase specifies the distance into the dash pattern at which to start the dash.
     *  The elements of both the dash array and the dash phase are expressed in user space units.
     *  <pre>
     *  Examples of line dash patterns:
     *
     *      "[Array] Phase"     Appearance          Description
     *      _______________     _________________   ____________________________________
     *
     *      "[] 0"              -----------------   Solid line
     *      "[3] 0"             ---   ---   ---     3 units on, 3 units off, ...
     *      "[2] 1"             -  --  --  --  --   1 on, 2 off, 2 on, 2 off, ...
     *      "[2 1] 0"           -- -- -- -- -- --   2 on, 1 off, 2 on, 1 off, ...
     *      "[3 5] 6"             ---     ---       2 off, 3 on, 5 off, 3 on, 5 off, ...
     *      "[2 3] 11"          -   --   --   --    1 on, 3 off, 2 on, 3 off, 2 on, ...
     *  </pre>
     *
     *  @param pattern the line dash pattern.
     */
    public void setPattern(String pattern) {
        this.pattern = pattern;
    }


    /**
     *  Sets the private fillShape variable.
     *  If the value of fillShape is true - the box is filled with the current brush color.
     *
     *  @param fillShape the value used to set the private fillShape variable.
     */
    public void setFillShape(boolean fillShape) {
        this.fillShape = fillShape;
    }


    /**
     *  Places this box in the another box.
     *
     *  @param box the other box.
     *  @param xOffset the x offset from the top left corner of the box.
     *  @param yOffset the y offset from the top left corner of the box.
     */
    public void placeIn(Box box, double xOffset, double yOffset) {
        placeIn(box, (float) xOffset, (float) yOffset);
    }


    /**
     *  Places this box in the another box.
     *
     *  @param box the other box.
     *  @param xOffset the x offset from the top left corner of the box.
     *  @param yOffset the y offset from the top left corner of the box.
     */
    public void placeIn(Box box, float xOffset, float yOffset) {
        this.x = box.x + xOffset;
        this.y = box.y + yOffset;
    }


    /**
     *  Scales this box by the spacified factor.
     *
     *  @param factor the factor used to scale the box.
     */
    public void scaleBy(double factor) {
        scaleBy((float) factor);
    }


    /**
     *  Scales this box by the spacified factor.
     *
     *  @param factor the factor used to scale the box.
     */
    public void scaleBy(float factor) {
        this.x *= factor;
        this.y *= factor;
    }


    /**
     *  Draws this box on the specified page.
     *
     *  @param page the page to draw this box on.
     *  @return x and y coordinates of the bottom right corner of this component.
     *  @throws Exception  If an input or output exception occurred
     */
    public float[] drawOn(Page page) throws Exception {
        page.addBMC(StructElem.SPAN, language, altDescription, actualText);
        page.setPenWidth(width);
        page.setLinePattern(pattern);
        if (fillShape) {
            page.setBrushColor(color);
        }
        else {
            page.setPenColor(color);
        }
        page.moveTo(x, y);
        page.lineTo(x + w, y);
        page.lineTo(x + w, y + h);
        page.lineTo(x, y + h);
        if (fillShape) {
            page.fillPath();
        }
        else {
            page.closePath();
        }
        page.addEMC();

        if (uri != null || key != null) {
            page.addAnnotation(new Annotation(
                    uri,
                    key,    // The destination name
                    x,
                    y,
                    x + w,
                    y + h,
                    language,
                    altDescription,
                    actualText));
        }

        return new float[] {x + w, y + h + width};
    }

}   // End of Box.java

File: src/box.go
package pdfjet

/**
 * box.go
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */

import (
	"github.com/edragoev1/pdfjet/src/color"
	"github.com/edragoev1/pdfjet/src/operator"
	"github.com/edragoev1/pdfjet/src/single"
	"github.com/edragoev1/pdfjet/src/structtype"
)

// Box is used to create rectangular boxes on a page.
// Also used to for layout purposes. See the placeIn method in the Image and TextLine classes.
type Box struct {
	x, y, w, h     float32
	r              float32
	color          int32
	width          float32
	pattern        string
	fillShape      bool
	uri            string
	key            string
	language       string
	altDescription string
	actualText     string
	structureType  string
}

// NewBox creates new Box object.
func NewBox() *Box {
	box := new(Box)
	box.color = color.Black
	box.width = 0.0
	box.pattern = "[] 0"
	box.altDescription = single.Space
	box.actualText = single.Space
	box.structureType = structtype.P
	return box
}

// NewBoxAt creates a box object.
// @param x the x coordinate of the top left corner of this box when drawn on the page.
// @param y the y coordinate of the top left corner of this box when drawn on the page.
// @param w the width of this box.
// @param h the height of this box.
func NewBoxAt(x, y, w, h float32) *Box {
	box := NewBox()
	box.x = x
	box.y = y
	box.w = w
	box.h = h
	return box
}

// SetLocation sets the location of this box on the page.
// @param x the x coordinate of the top left corner of this box when drawn on the page.
// @param y the y coordinate of the top left corner of this box when drawn on the page.
func (box *Box) SetLocation(x, y float32) *Box {
	box.x = x
	box.y = y
	return box
}

// SetSize sets the size of this box.
// @param w the width of this box.
// @param h the height of this box.
func (box *Box) SetSize(w, h float32) {
	box.w = w
	box.h = h
}

// SetColor sets the color for this box.
// @param color the color specified as an integer.
func (box *Box) SetColor(color int32) {
	box.color = color
}

// SetLineWidth sets the width of this line.
// @param width the width.
func (box *Box) SetLineWidth(width float32) {
	box.width = width
}

// SetCornerRadius sets the corner radius.
// @param width the width.
func (box *Box) SetCornerRadius(r float32) {
	box.r = r
}

// SetURIAction sets the URI for the "click box" action.
// @param uri the URI
func (box *Box) SetURIAction(uri string) {
	box.uri = uri
}

// SetGoToAction sets the destination key for the action.
// @param key the destination name.
func (box *Box) SetGoToAction(key string) {
	box.key = key
}

// SetAltDescription sets the alternate description of this box.
// @param altDescription the alternate description of the box.
// @return this Box.
func (box *Box) SetAltDescription(altDescription string) *Box {
	box.altDescription = altDescription
	return box
}

// SetActualText sets the actual text for this box.
// @param actualText the actual text for the box.
// @return this Box.
func (box *Box) SetActualText(actualText string) *Box {
	box.actualText = actualText
	return box
}

// SetStructureType sets the type of the structure.
func (box *Box) SetStructureType(structureType string) *Box {
	box.structureType = structureType
	return box
}

// SetPattern sets the line dash pattern that controls the pattern of dashes and gaps used to stroke paths.
// It is specified by a dash array and a dash phase.
// The elements of the dash array are positive numbers that specify the lengths of
// alternating dashes and gaps.
// The dash phase specifies the distance into the dash pattern at which to start the dash.
// The elements of both the dash array and the dash phase are expressed in user space units.
// <pre>
// Examples of line dash patterns:
//
//	"[Array] Phase"     Appearance          Description
//	_______________     _________________   ____________________________________
//
//	"[] 0"              -----------------   Solid line
//	"[3] 0"             ---   ---   ---     3 units on, 3 units off, ...
//	"[2] 1"             -  --  --  --  --   1 on, 2 off, 2 on, 2 off, ...
//	"[2 1] 0"           -- -- -- -- -- --   2 on, 1 off, 2 on, 1 off, ...
//	"[3 5] 6"             ---     ---       2 off, 3 on, 5 off, 3 on, 5 off, ...
//	"[2 3] 11"          -   --   --   --    1 on, 3 off, 2 on, 3 off, 2 on, ...
//
// </pre>
//
// @param pattern the line dash pattern.
func (box *Box) SetPattern(pattern string) {
	box.pattern = pattern
}

// SetFillShape sets the private fillShape variable.
// If the value of fillShape is true - the box is filled with the current brushColor color.
// @param fillShape the value used to set the private fillShape variable.
func (box *Box) SetFillShape(fillShape bool) {
	box.fillShape = fillShape
}

// PlaceIn places this box in the another box.
// @param box the other box.
// @param xOffset the x offset from the top left corner of the box.
// @param yOffset the y offset from the top left corner of the box.
func (box *Box) PlaceIn(box2 *Box, xOffset, yOffset float32) {
	box.x = box2.x + xOffset
	box.y = box2.y + yOffset
}

// ScaleBy scales this box by the specified factor.
// @param factor the factor used to scale the box.
func (box *Box) ScaleBy(factor float32) {
	box.x *= factor
	box.y *= factor
}

// DrawOn draws this box on the specified page.
// @param page the page to draw this box on.
// @return x and y coordinates of the bottom right corner of this component.
func (box *Box) DrawOn(page *Page) []float32 {
	const k float32 = 0.5517

	page.AddBMC(box.structureType, box.language, box.actualText, box.altDescription)
	if box.r == 0.0 {
		page.MoveTo(box.x, box.y)
		page.LineTo(box.x+box.w, box.y)
		page.LineTo(box.x+box.w, box.y+box.h)
		page.LineTo(box.x, box.y+box.h)
		if box.fillShape {
			page.SetBrushColor(box.color)
			page.FillPath()
		} else {
			page.SetPenWidth(box.width)
			page.SetPenColor(box.color)
			page.SetStrokeDashPattern(box.pattern)
			page.ClosePath()
		}
	} else {
		page.SetPenWidth(box.width)
		page.SetPenColor(box.color)
		page.SetStrokeDashPattern(box.pattern)

		points := make([]*Point, 0)
		points = append(points, NewPoint(box.x+box.r, box.y))
		points = append(points, NewPoint((box.x+box.w)-box.r, box.y))
		points = append(points, NewControlPointC(((box.x+box.w)-box.r)+box.r*k, box.y))
		points = append(points, NewControlPointC(box.x+box.w, (box.y+box.r)-box.r*k))
		points = append(points, NewPoint(box.x+box.w, box.y+box.r))
		points = append(points, NewPoint(box.x+box.w, (box.y+box.h)-box.r))
		points = append(points, NewControlPointC(box.x+box.w, ((box.y+box.h)-box.r)+box.r*k))
		points = append(points, NewControlPointC(((box.x+box.w)-box.r)+box.r*k, box.y+box.h))
		points = append(points, NewPoint((box.x+box.w)-box.r, box.y+box.h))
		points = append(points, NewPoint(box.x+box.r, box.y+box.h))
		points = append(points, NewControlPointC((box.x+box.r)-box.r*k, box.y+box.h))
		points = append(points, NewControlPointC(box.x, ((box.y+box.h)-box.r)+box.r*k))
		points = append(points, NewPoint(box.x, (box.y+box.h)-box.r))
		points = append(points, NewPoint(box.x, box.y+box.r))
		points = append(points, NewControlPointC(box.x, (box.y+box.r)-box.r*k))
		points = append(points, NewControlPointC((box.x+box.r)-box.r*k, box.y))
		points = append(points, NewPoint(box.x+box.r, box.y))

		page.DrawPath(points, operator.Stroke)
	}
	page.AddEMC()

	if box.uri != "" || box.key != "" {
		page.AddAnnotation(NewAnnotation(
			"Link",
			box.x,
			box.y,
			box.x+box.w,
			box.y+box.h,
			nil,
			nil,
			0.0,
			"",
			"",
			box.uri,
			box.key, // The destination name
			box.language,
			box.actualText,
			box.altDescription))
	}

	return []float32{box.x + box.w, box.y + box.h}
}
